numpy/core/src/multiarray/arrayobject.c

```diff
@@ -1626,4 +1682,7 @@ NPY_NO_EXPORT PyTypeObject PyArray_Type = {
     0,                                          /* tp_weaklist */
     0,                                          /* tp_del */
     0,                                          /* tp_version_tag */
+#if PY_VERSION_HEX >= 0x03040000
```

Comment:
Any other types you can think of that might need this? What is the determining feature?

Reply:
Py_INCREF(self) in `tp_dealloc` for the type. I didn't check if that occurs in other numpy extension types.

Reply:
I checked and I think only ndarray does this. (There are increfs in arraydescr_dealloc but they are in some obscure error handling path that also prints stuff to stderr --- looks bogus and probably that code is not reached in practice.)